kong/clustering/rpc/queue.lua

```diff
@@ -0,0 +1,71 @@
+local semaphore = require("ngx.semaphore")
+local table_new = require("table.new")
+
+
+local assert = assert
+local setmetatable = setmetatable
+local math_min = math.min
+
+
+local _M = {}
+local _MT = { __index = _M, }
+
+
+local DEFAULT_QUEUE_LEN = 128
+
+
+function _M.new(max_len)
+  local self = {
+    semaphore = assert(semaphore.new()),
+    max = max_len,
+
+    elts = table_new(math_min(max_len, DEFAULT_QUEUE_LEN), 0),
+    first = 0,
+    last = -1,
+  }
+
+  return setmetatable(self, _MT)
+end
+
+
+function _M:push(item)
+  local last = self.last
+
+  if last - self.first + 1 >= self.max then
+    return nil, "queue overflow"
+  end
+
+  last = last + 1
+  self.last = last
+  self.elts[last] = item
+
+  self.semaphore:post()
+
+  return true
+end
+
+
+function _M:pop(timeout)
+  local ok, err = self.semaphore:wait(timeout)
```

Comment:
should we give `timeout` a default value like `5`? then we can invoke it clearly.

Reply:
That would be less explicit because different place use different timeout value, I would rather this be specified manually by the caller.